packer/provisioner.go

```diff
@@ -162,6 +162,32 @@ func (p *PausedProvisioner) Provision(ctx context.Context, ui Ui, comm Communica
 	return p.Provisioner.Provision(ctx, ui, comm, generatedData)
 }
 
+// RetriedProvisioner is a Provisioner implementation that retries
+// the provisioner whenever there's an error.
+type RetriedProvisioner struct {
+	MaxRetries  int
+	Provisioner Provisioner
+}
+
+func (r *RetriedProvisioner) ConfigSpec() hcldec.ObjectSpec { return r.ConfigSpec() }
+func (r *RetriedProvisioner) FlatConfig() interface{}       { return r.FlatConfig() }
+func (r *RetriedProvisioner) Prepare(raws ...interface{}) error {
+	return r.Provisioner.Prepare(raws...)
+}
+
+func (r *RetriedProvisioner) Provision(ctx context.Context, ui Ui, comm Communicator, generatedData map[string]interface{}) error {
+	err := r.Provisioner.Provision(ctx, ui, comm, generatedData)
+
+	retries := 0
+	for err != nil && retries < r.MaxRetries {
+		ui.Say("Retrying provisioner")
+		err = r.Provisioner.Provision(ctx, ui, comm, generatedData)
+		retries++
+	}
```

Comment:
Ah I just thought about this, we should probably not retry in case the context is done too. Also I think we should say that we retry only in the case we actually retry, and may be showing the numder, what do you think ?:

```suggestion
	leftTries := r.MaxRetries
	for ; leftTries >= 0 ; leftTries-- {
		if ctx.Err() != nil { // context was cancelled
			return ctx.Err()
		}
		err := r.Provisioner.Provision(ctx, ui, comm, generatedData)
		if err == nil {
			return nil
		}
		if leftTries > 0 {
			ui.Say(fmt.Sprintf("Provisioner failed with %q, retrying with %d trie(s) left", err, leftTries))
		}
	}
	ui.Say("retry limit reached.")
```


 this should be also handled by most provisioners but not if they are say from a plugins that doesn't support context

Reply:
Good cath!